ports/geckolib/gecko_bindings/sugar/refptr.rs

```diff
@@ -1,23 +1,33 @@
 use std::marker::PhantomData;
 use std::mem::{forget, transmute};
-use std::ptr::{read, null};
+use std::ptr;
 use std::sync::Arc;
 
 /// Helper trait for conversions between FFI Strong/Borrowed types and Arcs
-pub unsafe trait RefCounted where Self: Sized {
+///
+/// Should be implemented by types which are passed over FFI as Arcs
+/// via Strong and Borrowed
+pub unsafe trait HasArcFFI where Self: Sized {
     /// Gecko's name for the type
-    /// This is equivalent to ArcBox<Self>
+    /// This is equivalent to ArcInner<Self>
     type FFIType: Sized;
+
+    /// Given a non-null borrowed FFI reference, this produces a temporary
+    /// Arc which is borrowed by the given closure and used.
+    /// Panics on null.
     fn with<F, Output>(raw: Borrowed<Self::FFIType>, cb: F) -> Output
```

Comment:
I don't like the duplication between with and maybe_with. Perhaps write it something like the following?

```rust
fn with<F, Output>(raw: Borrowed<Self::FFIType>, cb: F) -> Output where F: FnOnce(&Arc<Self>) -> Output {
    with_opt(move |a| cb(a.unwrap()))
}

fn with_opt<F, Output>(raw: Borrowed<Self::FFIType>, cb: F) -> Output where F: FnOnce(Option<&Arc<Self>>) -> Output {
    let owned = if raw.is_null() {
        None
    } else {
        Some(unsafe { Self::borrowed_into(raw) })
    };

    let result = cb(owned.as_ref());
    // This is not unwind safe. Do not use in builds without panic=abort enabled.
    forget(owned);
    result
}
```